#! /bin/sh /usr/share/dpatch/dpatch-run
## 50_fix_hold_state.dpatch by  <root@localhost>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix for #391033, thanks John Eaton

@DPATCH@
diff -urNad octave2.9-2.9.9~/scripts/plot/__clear_plot__.m octave2.9-2.9.9/scripts/plot/__clear_plot__.m
--- octave2.9-2.9.9~/scripts/plot/__clear_plot__.m	1970-01-01 01:00:00.000000000 +0100
+++ octave2.9-2.9.9/scripts/plot/__clear_plot__.m	2006-10-06 16:20:23.900513248 +0200
@@ -0,0 +1,42 @@
+## Copyright (C) 2006 John W. Eaton
+##
+## This file is part of Octave.
+##
+## Octave is free software; you can redistribute it and/or modify it
+## under the terms of the GNU General Public License as published by
+## the Free Software Foundation; either version 2, or (at your option)
+## any later version.
+##
+## Octave is distributed in the hope that it will be useful, but
+## WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+## General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with Octave; see the file COPYING.  If not, write to the Free
+## Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+## 02110-1301, USA.
+
+function __clear_plot__ (cmd, sep, clear_data)
+
+  __plot_globals__
+
+  if (nargin < 3)
+    clear_data = true;
+    if (nargin < 2)
+      sep = "";
+      if (nargin < 1)
+       cmd = "";
+      endif
+    endif
+  endif
+
+  __plot_command__{__current_figure__}{__multiplot_xi__,__multiplot_yi__} = cmd;
+  __plot_command_sep__ = sep;
+
+  if (clear_data)
+    __plot_data__{__current_figure__}{__multiplot_xi__,__multiplot_yi__} = [];
+    __plot_data_offset__{__current_figure__}(__multiplot_xi__,__multiplot_yi__) = 1;
+  endif
+
+endfunction
diff -urNad octave2.9-2.9.9~/scripts/plot/__setup_plot__.m octave2.9-2.9.9/scripts/plot/__setup_plot__.m
--- octave2.9-2.9.9~/scripts/plot/__setup_plot__.m	2006-09-26 23:16:52.000000000 +0200
+++ octave2.9-2.9.9/scripts/plot/__setup_plot__.m	2006-10-06 16:20:23.900513248 +0200
@@ -22,16 +22,20 @@
   __plot_globals__
 
   if (ishold ())
-    if (isempty (__plot_command__{__current_figure__}{__multiplot_xi__,__multiplot_yi__}))
-      __plot_command__{__current_figure__}{__multiplot_xi__,__multiplot_yi__} = plotcmd;
-      __plot_command_sep__ = "";
+    cmd = __plot_command__{__current_figure__}{__multiplot_xi__,__multiplot_yi__};
+    if (isempty (cmd))
+      cmd = plotcmd;
+      sep = "";
     else
-      __plot_command_sep__ = ",\\\n";
+      sep = ",\\\n";
     endif
+    clear_data = false;
   else
-    __plot_command__{__current_figure__}{__multiplot_xi__,__multiplot_yi__} = plotcmd;
-    __plot_command_sep__ = "";
-    __plot_data__{__current_figure__}{__multiplot_xi__,__multiplot_yi__} = [];
-    __plot_data_offset__{__current_figure__}(__multiplot_xi__,__multiplot_yi__) = 1;
+    cmd = plotcmd;
+    sep = "";
+    clear_data = true;
   endif
+
+  __clear_plot__ (cmd, sep, clear_data);
+
 endfunction
diff -urNad octave2.9-2.9.9~/src/DLD-FUNCTIONS/__gnuplot_raw__.l octave2.9-2.9.9/src/DLD-FUNCTIONS/__gnuplot_raw__.l
--- octave2.9-2.9.9~/src/DLD-FUNCTIONS/__gnuplot_raw__.l	2006-08-23 21:54:47.000000000 +0200
+++ octave2.9-2.9.9/src/DLD-FUNCTIONS/__gnuplot_raw__.l	2006-10-06 16:20:23.901513036 +0200
@@ -1601,9 +1601,11 @@
   octave_value_list args;
 
   args(0) = "off";
-
   feval ("hold", args);
 
+  args.resize (0);
+  feval ("__clear_plot__", args);
+
   return octave_value_list ();
 }
 
